java/src/main/java/actorworkflow/CustomerLoyaltyWorkflowImpl.java

```diff
@@ -0,0 +1,147 @@
+
+        if (info.getContinuedExecutionRunId().isEmpty()) {
+            String tier = this.customer.status().name();
+            activities.sendEmail(EmailStrings.EMAIL_WELCOME.formatted(tier));
+        }
+
+        // block on everything
+        while (true) {
+            Workflow.await(() -> !accountActive || info.getHistoryLength() > Shared.HISTORY_THRESHOLD);
+            if (accountActive) {
+                logger.info("Account still active, history size crossed limit; continuing-as-new.");
+                Workflow.continueAsNew(this.customer);
+            } else {
+                logger.info("Account canceled. Closing workflow.");
+                return "Done";
+            }
+        }
+    }
+
+    @Override
+    public void addLoyaltyPoints(int pointsToAdd) {
+        customer = customer.withPoints(customer.loyaltyPoints() + pointsToAdd);
+        logger.info("Added {} points to customer. Total loyalty points now {}.",
+                pointsToAdd, customer.loyaltyPoints());
+
+        StatusTier tierToPromoteTo = StatusTier.getMaxTier(customer.loyaltyPoints());
+
+        if (customer.status().minimumPoints() < tierToPromoteTo.minimumPoints()) {
+            logger.info("Promoting customer!");
+            customer = customer.withStatus(tierToPromoteTo);
+            activities.sendEmail(EmailStrings.EMAIL_PROMOTED.formatted(tierToPromoteTo.name()));
+        }
+    }
+
+    @Override
+    public void inviteGuest(Customer guest) {
+        logger.info("Checking to see if customer can invite guests.");
+        if (Customer.canAddGuest(customer)) {
+            logger.info("Customer is allowed to invite guests; attempting to start workflow for guest ID {}.",
+                    guest.customerId());
+            customer.guests().add(guest);
+
+            StatusTier guestMinStatus = StatusTier.previous(customer.status());
+            guest = guest.withStatus(guestMinStatus);
+
+            String guestWorkflowId = Shared.WORKFLOW_ID_FORMAT.formatted(guest.customerId());
+            ChildWorkflowOptions options =
+                    ChildWorkflowOptions.newBuilder()
+                            .setWorkflowId(guestWorkflowId)
+                            .setParentClosePolicy(ParentClosePolicy.PARENT_CLOSE_POLICY_ABANDON)
+                            .build();
+
+            CustomerLoyaltyWorkflow child = Workflow.newChildWorkflowStub(CustomerLoyaltyWorkflow.class, options);
+
+            boolean alreadyStarted = false;
+            try {
+                Promise<WorkflowExecution> childExecution = Workflow.getWorkflowExecution(child);
+                Async.procedure(child::customerLoyalty, guest);
+
+                // Wait for child to start
+                childExecution.get();
+            } catch (WorkflowExecutionAlreadyStarted e) {
+                logger.info("Guest customer workflow already started and is a direct child.");
+                alreadyStarted = true;
+            } catch (Exception e) {
```

Comment:
dont think should catch generic exception. irc this should always be ChildWorkflowFailure which would have the cause of WorkflowExecutionAlreadyStarted

Reply:
Updated!